### PR TITLE
Add support for `if else` expressions

### LINKED ISSUE #273: Support new conditional operator (ternary)
Delphi 13 added a new if-expression syntax similar to `?:` from other languages ([documentation](https://docwiki.embarcadero.com/RADStudio/Florence/en/Conditional_Operators_(Delphi))).

Since this operator is intended for terse code, pasfmt should try to format such expressions inline as much as possible.

Examples of the (proposed) desired formatting include
```delphi
A := if B then C else D;

Foo(if B then C else D);

if (if B then C else D) > 10 then
  Foo
else
  Bar;
```

In cases where a line break is forced, or it doesn't fit on one line, we should revert to the existing formatting for if-then-else **statements**, with continuations appropriate for the context:
```delphi
A := //
    if B then
        C
    else
        D;

Foo( //
    if B then
        C
    else
        D;
);

if ( //
    if B then
        C
    else
        D)
    > 10 then
  Foo
else
  Bar;
```

### PR DESCRIPTION
**Pull request by jgardn3r**
This PR closes #273.

There are some commits from the compound operator PR that will disappear after it is merged.
Adding a feature like this requires changes to contexts, rules, and the overall understanding of the roles `if` and `else` play in the context of a line.

### CODE REVIEW

File: core/datatests/generators/optimising_line_formatter.rs [hunk 1]
```diff
@@ -432,6 +432,31 @@ mod comments {
                             and FFFFFFFFFFFF then
                       ;
                 ",
+                ternary = "
```
Comment (fourls) [inline core/datatests/generators/optimising_line_formatter.rs:435]: Should we include ternaries with anonymous methods as the expressions too?

File: core/src/rules/optimising_line_formatter/contexts.rs
Comment (fourls) [inline core/src/rules/optimising_line_formatter/contexts.rs]: Update vague comment to explain:
- is useless because...?
- This removes single-token contexts, but it also removes comma lists and assignments with no end token. Why?

File: core/CHANGELOG.md [hunk 2]
```diff
@@ -7,6 +7,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## Unreleased
 
+### Added
+
+- Added support for `if else` ternary expressions
```
Comment (fourls) [inline core/CHANGELOG.md]: Needs full stop

File: CHANGELOG.md [hunk 3]
```diff
@@ -7,6 +7,10 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 
 ## [Unreleased]
 
+### Added
+
+- Added support for `if else` ternary expressions
```
Comment (fourls) [inline CHANGELOG.md]: Needs full stop

File: core/src/rules/optimising_line_formatter/requirements.rs [hunk 4]
```diff
@@ -278,7 +278,20 @@ impl InternalOptimisingLineFormatter<'_, '_> {
                 .get_last_context(CT::ControlFlowBegin)
                 .map(|(_, data)| data.is_child_broken)
                 .if_else_or_default(DR::MustBreak, DR::Indifferent),
-            (_, Some(TT::Keyword(KK::Else))) => DR::MustBreak,
+            (_, Some(TT::Keyword(KK::Else))) => contexts_data
+                .get_last_context(context_matches!(CT::IfElse))
+                .map(|(_, data)| data.is_broken | data.is_child_broken)
+                .if_else_or(DR::MustBreak, DR::MustNotBreak, DR::MustBreak),
+            (Some(TT::Keyword(KK::Then)), Some(TT::Keyword(KK::If))) => DR::MustBreak,
```
Comment (fourls) [inline core/src/rules/optimising_line_formatter/requirements.rs:285]: Just to confirm my own understanding here, these don't apply to if _statements_, only if _expressions_, because this code only triggers for a `then` immediately followed by an `if` on the same logical line?

File: core/src/rules/optimising_line_formatter/contexts.rs [hunk 5]
```diff
@@ -2225,6 +2304,51 @@ mod tests {
...
+                1 AssignRHS          ^----------------------------------$
+                0 IfElse             ^--------------------------$--------
+                1 EGuardClause          ^-----$
+                1 Precedence(4)         ^-----$
+                1 Subject                       ^----------$
+                1 Precedence(3)                      ^-----$
+                1 Subject                                    ^----------$
+                1 Precedence(3)                                   ^-----$
+            "},
+            nested = {"
```
Comment (fourls) [inline core/src/rules/optimising_line_formatter/contexts.rs]: Let's add a test for nested in the `else`.

File: core/src/rules/optimising_line_formatter/contexts.rs [hunk 6]
```diff
@@ -1077,12 +1112,48 @@ impl<'a> LineFormattingContexts<'a> {
...
                 TT::Keyword(KK::If | KK::While | KK::With | KK::On) => {
                     contexts.push_utility((CT::ControlFlowBegin, 0));
                     contexts.push(CT::ControlFlow);
                 }
                 TT::Keyword(KK::Else) => {
-                    contexts.pop_until(CT::ControlFlowBegin);
+                    if let Some(CT::IfElse) =
+                        contexts.pop_until(context_matches!(CT::ControlFlowBegin | CT::IfElse))
+                    {
+                        // This is to ensure nested if-else expressions are popped of the stack.
```
Comment (fourls) [inline core/src/rules/optimising_line_formatter/contexts.rs]:
```suggestion
                        // This is to ensure nested if-else expressions are popped off the stack.
```

File: core/src/rules/optimising_line_formatter/contexts.rs [hunk 7]
```diff
@@ -895,7 +911,26 @@ impl<'a> LineFormattingContexts<'a> {
                             contexts.push_expression();
                         }
                         TT::Keyword(KK::If | KK::While | KK::On | KK::Until | KK::Case) => {
-                            contexts.push(CT::GuardClause);
+                            contexts.push(
+                                if prev_token_types
+                                    .iter()
+                                    .filter(|tt| !tt.is_comment_or_compiler_directive())
```
Comment (fourls) [inline core/src/rules/optimising_line_formatter/contexts.rs]: Should this be `is_comment_or_directive()`?

File: core/src/rules/optimising_line_formatter/contexts.rs [hunk 8]
```diff
@@ -895,7 +911,26 @@ impl<'a> LineFormattingContexts<'a> {
...
+                            );
+                            contexts.push_expression();
+                        }
+                        TT::Keyword(KK::Then) => {
+                            // if-else expressions
+                            contexts.push_expression();
+                        }
+                        TT::Keyword(KK::Else) => {
+                            // if-else expressions
                             contexts.push_expression();
```
Comment (fourls) [inline core/src/rules/optimising_line_formatter/contexts.rs:934]: Wouldn't these also apply for if-else statements? Could that cause any issues? Or not a big deal because these would be the last tokens on the logical line?

**Review by fourls — CHANGES_REQUESTED** [review]
I have played around with it a bit and I'm mostly happy with the rule as it is. Let's add some more complex examples in the tests, especially with anonymous methods and with nesting in the `else` block, and I think we should have most of the edge cases.

Also there is a commit message typo: Convert get_operator_*precedence* to a trait routine

File: core/datatests/generators/optimising_line_formatter.rs (hunk 1 above)
Comment (jgardn3r) [reply core/datatests/generators/optimising_line_formatter.rs:435]: Why not, ill stick some in the `expressions::ternary` module.

File: core/src/rules/optimising_line_formatter/requirements.rs (hunk 4 above)
Comment (jgardn3r) [reply core/src/rules/optimising_line_formatter/requirements.rs:285]: That is correct. If statements' statements are child lines.

File: core/src/rules/optimising_line_formatter/contexts.rs (hunk 7 above)
Comment (jgardn3r) [reply core/src/rules/optimising_line_formatter/contexts.rs]: I don't think a conditional directive in that location is currently possible. But for the sake of being defensive to future updates, seems like a good idea.

File: core/src/rules/optimising_line_formatter/contexts.rs (hunk 8 above)
Comment (jgardn3r) [reply core/src/rules/optimising_line_formatter/contexts.rs:934]: They would apply to it-else statements too. I don't think it is relevant for the reason you say. On the statement line, `then` and `else` are next to each other. 

As such, the line that looks like the following will subsequently pop all the contexts that were just added.
```rust
if let Some(CT::IfElse) =
    contexts.pop_until(context_matches!(CT::ControlFlowBegin | CT::IfElse))
```

You might argue that it's doing more work than it should, but due to the nature of Delphi and its grammar, that is how the context construction is currently laid out.
